python/models/hierarchical_filtering_proxy_model.py

```diff
@@ -0,0 +1,158 @@
+# 
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit 
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your 
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights 
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+"""
+Proxy model that provides efficient hierarhcical filtering of a tree-based source model
+"""
+
+import sgtk
+from sgtk.platform.qt import QtCore, QtGui
+
+class HierarchicalFilteringProxyModel(QtGui.QSortFilterProxyModel):
+    """
+    Inherited from a QSortFilterProxyModel, this class implements filtering across all 
+    levels of the hierarchy in a hierarchical (tree-based) model
+    """
+    
+    def __init__(self, parent=None):
+        """
+        Construction
+        :param parent:    The parent QObject to use for this instance
+        """
+        QtGui.QSortFilterProxyModel.__init__(self, parent)
+        
+        self._cached_regexp = None
+        self._filter_dirty = True
+        self._accepted_cache = {}
+        self._child_accepted_cache = {}
+
+    def _is_item_accepted(self, item, parent_accepted):
+        """
+        Override this method to decide if the specified item should be accepted or 
+        not by the filter.
+        
+        This should be overridden instead of filterAcceptsRow in derived classes
+        
+        :param item:            The QStandardItem item
+        :param parent_accepted: True if a parent item has been accepted by the filter
+        :returns:               True if this item should be accepted, otherwise False
+        """
+        raise NotImplementedError()
+
+    # -------------------------------------------------------------------------------
+    # -------------------------------------------------------------------------------
+
+    def invalidateFilter(self):
+        """
+        Overriden base class method used to invalidate the current filter.
+        """
+        self._filter_dirty = True
+        # call through to the base class:
+        QtGui.QSortFilterProxyModel.invalidateFilter(self)
+
+    def filterAcceptsRow(self, src_row, src_parent_idx):
+        """
+        Overriden base class method used to determine if a row is accepted by the
+        current filter.
+        
+        This implementation checks both up and down the hierarchy to determine if
+        this row should be accepted.
+
+        :param src_row:         The row in the source model to filter
+        :param src_parent_idx:  The parent index in the source model to filter
+        :returns:               True if the row should be accepted by the filter, False
+                                otherwise
+        """
+        reg_exp = self.filterRegExp()
+        if self._filter_dirty or reg_exp != self._cached_regexp:
+            # clear the cache as the search filter has changed
+            self._accepted_cache = {}
```

Comment:
Not a big deal, but a _clear_cache method shared between this method and __init__ that takes the regex and the filter flag would have been nice. 4 things to keep track of when clearing the cache sounds like a maintenance nightmare when the cache is cleared more often or items are added to the cache.